docs/templates/tutorials/subclass-tuner.md

```diff
@@ -0,0 +1,132 @@
+            activation='relu',
+            padding='same')(x)
+
+        if hp.Choice('pooling' + str(i), ['max', 'avg']) == 'max':
+            x = tf.keras.layers.MaxPooling2D()(x)
+        else:
+            x = tf.keras.layers.AveragePooling2D()(x)
+
+        x = tf.keras.layers.BatchNormalization()(x)
+        x = tf.keras.layers.ReLU()(x)
+
+    if hp.Choice('global_pooling', ['max', 'avg']) == 'max':
+        x = tf.keras.layers.GlobalMaxPooling2D()(x)
+    else:
+        x = tf.keras.layers.GlobalAveragePooling2D()(x)
+    outputs = tf.keras.layers.Dense(10, activation='softmax')(x)
+
+    model = tf.keras.Model(inputs, outputs)
+
+    optimizer = hp.Choice('optimizer', ['adam', 'sgd'])
+    model.compile(optimizer, loss='sparse_categorical_crossentropy', metrics=['accuracy'])
+    return model
+
+
+class MyTuner(kt.Tuner):
+
+    def run_trial(self, trial, train_ds):
+        hp = trial.hyperparameters
+
+        # Hyperparameters can be added anywhere inside `run_trial`.
+        # When the first trial is run, they will take on their default values.
+        # Afterwards, they will be tuned by the `Oracle`.
+        train_ds = train_ds.batch(
+            hp.Int('batch_size', 32, 128, step=32, default=64))
+
+        model = self.hypermodel.build(trial.hyperparameters)
+        lr = hp.Float('learning_rate', 1e-4, 1e-2, sampling='log', default=1e-3)
+        optimizer = tf.keras.optimizers.Adam(lr)
+        epoch_loss_metric = tf.keras.metrics.Mean()
+
+        @tf.function
+        def run_train_step(data):
+            images = tf.dtypes.cast(data['image'], 'float32') / 255.
+            labels = data['label']
+            with tf.GradientTape() as tape:
+                logits = model(images)
+                loss = tf.keras.losses.sparse_categorical_crossentropy(
+                    labels, logits)
+                # Add any regularization losses.
+                if model.losses:
+                    loss += tf.math.add_n(model.losses)
+                gradients = tape.gradient(loss, model.trainable_variables)
+            optimizer.apply_gradients(zip(gradients, model.trainable_variables))
+            epoch_loss_metric.update_state(loss)
+            return loss
+
+        # `self.on_epoch_end` reports results to the `Oracle` and saves the
+        # current state of the Model. The other hooks called here only log values
+        # for display but can also be overridden. For use cases where there is no
+        # natural concept of epoch, you do not have to call any of these hooks. In
+        # this case you should instead call `self.oracle.update_trial` and
+        # `self.oracle.save_model` manually.
+        for epoch in range(10):
+            print('Epoch: {}'.format(epoch))
+
+            self.on_epoch_begin(trial, model, epoch, logs={})
+            for batch, data in enumerate(train_ds):
+                self.on_batch_begin(trial, model, batch, logs={})
+                batch_loss = run_train_step(data).numpy()
```

Comment:
This is a scalar so prefer doing `float(run_train_step(data))` (more readable than `.numpy()`, and gives you a plain Python float).

Reply:
Done